### ISSUE TITLE
Directive controllerAs and scope

### ISSUE THREAD

**Issue by facultymatt** [johnpapa/angular-styleguide#536]
When using multiple directives that use the `controllerAs: 'vm'` syntax and do not have scope, they will conflict and `vm` will get overwritten. For example, the following directives will not work as expected:

```
angular
    .module('app')
    .directive('myExampleOne', myExampleOne);

function myExampleOne() {
    var directive = {
        templateUrl: 'app/feature/example.directiveOne.html',
        controllerAs: 'vm',
        bindToController: true 
    };
    return directive;
}

angular
    .module('app')
    .directive('myExampleTwo', myExampleTwo);

function myExampleTwo() {
    var directive = {
        templateUrl: 'app/feature/example.directiveTwo.html',
        controllerAs: 'vm',
        bindToController: true 
    };
    return directive;
}
```

A few questions here: 

1. Does it make sense to always use scope? `scope: {}` or even `scope: true` will fix this issue. In my experience this is good practice to avoid scope conflicts even sans `controllerAs`.
2. Should we not use `bindToController` in these examples?

## Related style 

https://github.com/johnpapa/angular-styleguide#style-y075

**Comment by lunks** [comment]
I'd say that it is a good practice to always create a new scope for directives.

**Comment by Jandalf** [comment]
using only isolated scopes, gives you more control over the dataflow of your application.
and yes since angular 1.4 you should always use bindToController with isolated scopes and controllerAs.

**Comment by lunks** [comment]
Any changes on angular 1.4 that makes this different than 1.3, @Jandalf?

**Comment by Jandalf** [comment]
@lunks yes bindToController supports now an object value to follow the angular 2 binding style

e.g.: `scope: {}, bindToController {param1: '='}`

see #543

**Comment by karlhorky** [comment]
Are isolated scopes not bad for composability (more directives on one element)? Since Angular 1.2 there is an error `Multiple Directive Resource Contention` when using multiple directives with isolated scopes. For instance:

```html
<input my-directive maxlength="20" ng-model="title">
```

https://docs.angularjs.org/error/$compile/multidir?p0=myDirective&p1=maxlength&p2=new%2Fisolatedundefinedcope&p3=undefined

**Comment by karlhorky** [comment]
I've opened #608 for an answer to this.

**Comment by karlhorky** [comment]
For this reason I would like an alternative answer to this question though - without using isolated scope. One thing I can imagine is to rename `controllerAs` based on the directive name.

```js
angular
    .module('app')
    .directive('myExampleOne', myExampleOne);

function myExampleOne() {
    var directive = {
        templateUrl: 'app/feature/example.directiveOne.html',
        controllerAs: 'myExampleOneVm',
        bindToController: true 
    };
    return directive;
}

angular
    .module('app')
    .directive('myExampleTwo', myExampleTwo);

function myExampleTwo() {
    var directive = {
        templateUrl: 'app/feature/example.directiveTwo.html',
        controllerAs: 'myExampleTwoVm',
        bindToController: true 
    };
    return directive;
}
```

This would still not solve nested directives of the same type though...

**Comment by johnpapa** [comment]
Thank you for the thoughtful suggestion! Since Angular 1 has reached end of life, the style guide has stabilized and we're no longer making structural changes. The guide remains a valuable reference as-is. Appreciate your contribution to the community! 🙏
